Update logging docs

noahfalk opened the pull request:
We've gotten feedback these docs are confusing and we are trying to make improvements:
- Tighten up the initial text to have fewer warnings, caveats and tips so we get to key concepts and examples faster
- Lead off with a minimal example to establish basic concepts. I tried experimenting with including more concepts in the first example like DI or source generated logging, but it felt both easier to understand and more concise starting with a low concept count example and then building out additional concepts and variations from there.
- Removed other extraneous C# features from the initial example as not all C# developers are familiar with them: BackgroundService, async/await, cancellation tokens.
- Added an explicit recommendation to use source generated logging. We probably should use it more in other examples, but this is a start.
- Tried to make the the article a little more balanced between hosted and un-hosted approaches. For the original ASP.NET docs it made sense to prioritize hosting and DI, but for the general purpose doc we should show this API can be equally applicable to all app models. I don't think it is even yet, but the balance is closer.


<!-- PREVIEW-TABLE-START -->

---

#### Internal previews

| 📄 File | 🔗 Preview link |
|:--|:--|
| [docs/core/extensions/logging.md](https://github.com/dotnet/docs/blob/6c28c51585c8982f138ad1d16d85f4d47102211d/docs/core/extensions/logging.md) | [Logging in C# and .NET](https://review.learn.microsoft.com/en-us/dotnet/core/extensions/logging?branch=pr-en-us-37506) |


<!-- PREVIEW-TABLE-END -->

noahfalk committed:
Update logging docs

We've gotten feedback these docs are confusing and we are trying to make improvements:
- Tighten up the initial text to have fewer warnings, caveats and tips so we get to key concepts and examples faster
- Lead off with a minimal example to establish basic concepts. I tried experimenting with including
more concepts in the first example like DI or source generated logging, but it felt both easier to
understand and more concise starting with a low concept count example and then building out
additional concepts and variations from there.
- Removed other extraneous C# features from the initial example as not all C# developers are familiar with them:
 BackgroundService, async/await, cancellation tokens.
- Added an explicit recommendation to use source generated logging. We probably should use it more in other
 examples, but this is a start.
- Tried to make the the article a little more balanced between hosted and un-hosted approaches. For the original
ASP.NET docs it made sense to prioritize hosting and DI, but for the general purpose doc we should show this API
 can be equally applicable to all app models. I don't think it is even yet, but the balance is closer.

noahfalk commented:
@IEvangelist @reyang @tarekgh @samsp-msft @davidfowl - could you guys take a look at this and let me know what you think?

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+
+- Creates an <xref:Microsoft.Extensions.Logging.ILoggerFactory>. The ILoggerFactory stores all the configuration that determines where log messages are sent. In this case
+we configured the console [logging provider](logging-providers.md) so that log messages are written to the console.
+- Creates an <xref:Microsoft.Extensions.Logging.ILogger> with a category named `Program`. The [category](#log-category) is a string will that be associated with each message logged
+by this ILogger object. It can be used to group log messages from the same class together when searching or filtering logs.
```
Use active voice, and correctly stylize references to code bits.

```suggestion
- Creates an <xref:Microsoft.Extensions.Logging.ILogger> with a category named "Program". The [category](#log-category) is a `string` that is associated with each message logged
by the `ILogger` object. It's used to group log messages from the same class (or category) together when searching or filtering logs.
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+we configured the console [logging provider](logging-providers.md) so that log messages are written to the console.
+- Creates an <xref:Microsoft.Extensions.Logging.ILogger> with a category named `Program`. The [category](#log-category) is a string will that be associated with each message logged
+by this ILogger object. It can be used to group log messages from the same class together when searching or filtering logs.
+- Calls <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogInformation%2A> to log a message at the `Information` level. The [log level](#log-level) indicates the severity of the logged event
+and can be used to filter out less important log messages. The log entry also includes a [message template](#log-message-template) `"Hello World! Logging is {description}."` and a key-value pair
+`description=fun`. The key name comes from the word inside the curly braces in the template and the value comes from the remaining method argument.
```
Corrected placeholder naming convention, and a bit of clean up.
 
```suggestion
- Calls <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogInformation%2A> to log a message at the `Information` level. The [log level](#log-level) indicates the severity of the logged event and is used to filter out less important log messages. The log entry also includes a [message template](#log-message-template) `"Hello World! Logging is {Description}."` and a key-value pair `Description = fun`. The key name (or placeholder) comes from the word inside the curly braces in the template and the value comes from the remaining method argument.
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
-> Starting with .NET 6, logging services no longer register the <xref:Microsoft.Extensions.Logging.ILogger> type. When using a logger, specify the generic-type alternative <xref:Microsoft.Extensions.Logging.ILogger%601> or register the `ILogger` with [dependency injection (DI)](dependency-injection.md).
+## Logging in a non-trivial app
+
+There are several changes you should consider making to the example above when logging in a less trivial scenario:
```
Avoid language like "above" or "below".

```suggestion
There are several changes you should consider making to the previous example when logging in a less trivial scenario:
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+
+There are several changes you should consider making to the example above when logging in a less trivial scenario:
+
+- If your application is using [Dependency Injection (DI)](dependency-injection.md) or a host such as ASP.NET's [WebApplication](/aspnet/core/fundamentals/minimal-apis/webapplication) or [Generic Host](generic-host.md) then
+you should use ILoggerFactory and ILogger objects from the DI container rather than creating them directly. See [Integration with DI and Hosts](#integration-with-hosts-and-dependency-injection) below for details.
```
When linking to additional information, use the following form:

```md
"For more information, see [link](other-content.md)."
```

```suggestion
- If your application is using [Dependency Injection (DI)](dependency-injection.md) or a host such as ASP.NET's [WebApplication](/aspnet/core/fundamentals/minimal-apis/webapplication) or [Generic Host](generic-host.md) then you should use `ILoggerFactory` and `ILogger` objects from their respective DI containers rather than creating them directly. For more information, see [Integration with DI and Hosts](#integration-with-hosts-and-dependency-injection).
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+- If your application is using [Dependency Injection (DI)](dependency-injection.md) or a host such as ASP.NET's [WebApplication](/aspnet/core/fundamentals/minimal-apis/webapplication) or [Generic Host](generic-host.md) then
+you should use ILoggerFactory and ILogger objects from the DI container rather than creating them directly. See [Integration with DI and Hosts](#integration-with-hosts-and-dependency-injection) below for details.
+
+- Logging [compile-time source generation](logger-message-generator.md) is usually a better alternative to ILogger extension methods like LogInformation(). It offers
+better performance, stronger typing, and avoids spreading string constants throughout your methods. The tradeoff is that using this technique requires a bit more code.
```
```suggestion
- Logging [compile-time source generation](logger-message-generator.md) is usually a better alternative to `ILogger` extension methods like `LogInformation`. Logging source generation offers better performance, stronger typing, and avoids spreading `string` constants throughout your methods. The tradeoff is that using this technique requires a bit more code.
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+
+:::code language="csharp" source="snippets/logging/getting-started-logger-message/Program.cs" range="5-13" highlight="5,8-9":::
+
+- The recommended practice for log category names is to use the name of the class that is creating the log message. This helps relate log messages back to the code which produced
+them and offers a good level of control when filtering logs. <xref:Microsoft.Extensions.Logging.LoggerFactoryExtensions.CreateLogger%2A> accepts a Type to make this naming easy to do.
```
```suggestion
- The recommended practice for log category names is to use the name of the class that's creating the log message. This helps relate log messages back to the code which produced them and offers a good level of control when filtering logs. <xref:Microsoft.Extensions.Logging.LoggerFactoryExtensions.CreateLogger%2A> accepts a `Type` to make this naming easy to do.
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+
+:::code language="csharp" source="snippets/logging/getting-started-type-category-name/Program.cs" range="3-11" highlight="6":::
+
+- If you don't use console logs as your sole production monitoring solution, add the [logging providers](logging-providers.md) you plan to use. For example you could use
+[OpenTelemetry](https://github.com/open-telemetry/opentelemetry-dotnet#getting-started) sending logs over the [OTLP protocol](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md#enable-log-exporter):
```
```suggestion
- If you don't use console logs as your sole production monitoring solution, add the [logging providers](logging-providers.md) you plan to use. For example, you could use [OpenTelemetry](https://github.com/open-telemetry/opentelemetry-dotnet#getting-started) to send logs over the [OTLP protocol](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md#enable-log-exporter):
```

reyang replied:
nit: OTLP's `P` stands for "protocol".

```suggestion
[OpenTelemetry](https://github.com/open-telemetry/opentelemetry-dotnet#getting-started) sending logs over the [OTLP (OpenTelemetry Protocol)](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md#enable-log-exporter):
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+
+## Integration with hosts and dependency injection
+
+If your application is using [Dependency Injection (DI)](dependency-injection.md) or a host such as ASP.NET's [WebApplication](/aspnet/core/fundamentals/minimal-apis/webapplication) or [Generic Host](generic-host.md) then
+you should use ILoggerFactory and ILogger objects from the DI container rather than creating them directly.
```
```suggestion
If your application is using [Dependency Injection (DI)](dependency-injection.md) or a host such as ASP.NET's [WebApplication](/aspnet/core/fundamentals/minimal-apis/webapplication) or [Generic Host](generic-host.md) then you should use `ILoggerFactory` and `ILogger` objects from the DI container rather than creating them directly.
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+:::code language="csharp" source="snippets/logging/aspnet-mvc/Controllers/HomeController.cs" range="33-37":::
 
-## Create logs
+<xref:Microsoft.Extensions.Logging.ILogger%601> derives from <xref:Microsoft.Extensions.Logging.ILogger> and indicates which category the `ILogger` object has. The DI container locates an `ILogger` with
+the correct category and supplies it as the constructor argument. If no `ILogger` with that category exists yet, the DI container automatically creates it from the `ILoggerFactory` that is in the container.
```
```suggestion
<xref:Microsoft.Extensions.Logging.ILogger%601> derives from <xref:Microsoft.Extensions.Logging.ILogger> and indicates which category the `ILogger` object has. The DI container locates an `ILogger` with the correct category and supplies it as the constructor argument. If no `ILogger` with that category exists yet, the DI container automatically creates it from the `ILoggerFactory` in the service provider.
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+### Host-provided ILoggerFactory
 
-The following example:
+Host builders initialize [default configuration](generic-host.md#default-builder-settings),
+then add a configured ILoggerFactory object to the host's DI container when the host is built. Before the host is built you can adjust the logging configuration
+via <xref:Microsoft.Extensions.Hosting.HostApplicationBuilder.Logging?displayProperty=nameWithType>, <xref:Microsoft.AspNetCore.Builder.WebApplicationBuilder.Logging?displayProperty=nameWithType>, or similar APIs on other
+hosts. Hosts also apply logging configuration from default configuration sources as *appsettings.json* and environment variables.
```
```suggestion
Host builders initialize [default configuration](generic-host.md#default-builder-settings),
then add a configured `ILoggerFactory` object to the host's DI container when the host is built. Before the host is built you can adjust the logging configuration via <xref:Microsoft.Extensions.Hosting.HostApplicationBuilder.Logging?displayProperty=nameWithType>, <xref:Microsoft.AspNetCore.Builder.WebApplicationBuilder.Logging?displayProperty=nameWithType>, or similar APIs on other hosts. Hosts also apply logging configuration from default configuration sources as _appsettings.json_ and environment variables. For more information, see [Configuration in .NET](configuration.md).
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
 
-- Creates a logger, `ILogger<Worker>`, which uses a log *category* of the fully qualified name of the type `Worker`. The log category is a string that is associated with each log.
-- Calls <xref:Microsoft.Extensions.Logging.LoggerExtensions.LogInformation%2A> to log at the `Information` level. The Log *level* indicates the severity of the logged event.
+This example customizes the ILoggerFactory provided by WebApplicationBuilder. It adds [OpenTelemetry](https://github.com/open-telemetry/opentelemetry-dotnet#getting-started) as a logging provider transmitting the logs over [OTLP](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md#enable-log-exporter):
```
```suggestion
This example customizes the `ILoggerFactory` provided by `WebApplicationBuilder`. It adds [OpenTelemetry](https://github.com/open-telemetry/opentelemetry-dotnet#getting-started) as a logging provider transmitting the logs over [OTLP](https://github.com/open-telemetry/opentelemetry-dotnet/blob/main/src/OpenTelemetry.Exporter.OpenTelemetryProtocol/README.md#enable-log-exporter):
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+:::code language="csharp" source="snippets/logging/aspnet-mvc/Program.cs" range="1-9" highlight="4":::
 
-[Levels](#log-level) and [categories](#log-category) are explained in more detail later in this article.
+### Creating an ILoggerFactory with DI
```
Gerunds.

```suggestion
### Create an ILoggerFactory with DI
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
-[Levels](#log-level) and [categories](#log-category) are explained in more detail later in this article.
+### Creating an ILoggerFactory with DI
+
+If you are using a DI container without a host, use <xref:Microsoft.Extensions.DependencyInjection.LoggingServiceCollectionExtensions.AddLogging%2A> to configure and add `ILoggerFactory` to the container.
```
```suggestion
If you're using a DI container without a host, use <xref:Microsoft.Extensions.DependencyInjection.LoggingServiceCollectionExtensions.AddLogging%2A> to configure and add `ILoggerFactory` to the container.
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+
+The preceding example:
+
+- Created a DI service container containing an ILoggerFactory configured to write to the console
```
```suggestion
- Created a DI service container containing an `ILoggerFactory` configured to write to the console
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
 ## Configure logging
 
-Logging configuration is commonly provided by the `Logging` section of *appsettings*.`{Environment}`*.json* files. The following *appsettings.Development.json* file is generated by the .NET Worker service templates:
+Logging configuration can be set in code or via external sources such as config files and environment variables. Using external configuration is beneficial when possible because
+it can be changed without rebuilding the application. However some tasks, such as setting logging providers, can only be configured from code.
```
Active voice, grammar, and remove extra lines.

```suggestion
Logging configuration is set in code or via external sources, such as, config files and environment variables. Using external configuration is beneficial when possible because it can be changed without rebuilding the application. However, some tasks, such as setting logging providers, can only be configured from code.
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+
+### Configure logging without code
+
+For apps that [use a host](#integration-with-hosts-and-dependency-injection), logging configuration is commonly provided by the `Logging` section of *appsettings*.`{Environment}`*.json* files.
+For apps that do not use a host, external configuration sources can either be [set up explicitly](configuration.md) or you can [configure logging in code](#configure-logging-with-code) instead.
```
```suggestion
For apps that [use a host](#integration-with-hosts-and-dependency-injection), logging configuration is commonly provided by the `"Logging"` section of _appsettings_`.{Environment}`_.json_ files. For apps that don't use a host, external configuration sources are [set up explicitly](configuration.md) or [configured in code](#configure-logging-with-code) instead.
```

IEvangelist commented on docs/core/extensions/snippets/logging/di-without-host/di-without-host.csproj:
```diff
@@ -0,0 +1,17 @@
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
+    <RootNamespace>di_without_host</RootNamespace>
```
```suggestion
    <RootNamespace>NonHost.Example</RootNamespace>
```

IEvangelist commented on docs/core/extensions/snippets/logging/getting-started-type-category-name/Program.cs:
```diff
@@ -0,0 +1,11 @@
+    {
+        using ILoggerFactory factory = LoggerFactory.Create(builder => builder.AddConsole());
+        ILogger logger = factory.CreateLogger<Program>();
+        logger.LogInformation("Hello World! Logging is {description}.", "fun");
```
```suggestion
        logger.LogInformation("Hello World! Logging is {Description}.", "fun");
```

reyang commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
+
+:::code language="csharp" source="snippets/logging/getting-started-logger-message/Program.cs" range="5-13" highlight="5,8-9":::
+
+- The recommended practice for log category names is to use the name of the class that is creating the log message. This helps relate log messages back to the code which produced
```
"name of the class" - is this the "fully" qualified name (e.g. `MyNamespace.MyClass`)? (I think the answer is YES)

noahfalk replied:
yes, the recommendation is to use the fully qualified name and that is the behavior that `CreateLogger<T>`() gives you.

IEvangelist commented on docs/core/extensions/snippets/logging/getting-started/Program.cs:
```diff
@@ -0,0 +1,5 @@
+
+using ILoggerFactory factory = LoggerFactory.Create(builder => builder.AddConsole());
+ILogger logger = factory.CreateLogger("Program");
+logger.LogInformation("Hello World! Logging is {description}.", "fun");
```
```suggestion
logger.LogInformation("Hello World! Logging is {Description}.", "fun");
```

IEvangelist commented on docs/core/extensions/snippets/logging/logging.sln:
```diff
@@ -19,6 +19,18 @@ Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "logger-message-generator",
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "getting-started-open-telemetry", "getting-started-open-telemetry\getting-started-open-telemetry.csproj", "{D711C802-982D-4712-B70D-A3179C4313CC}"
+EndProject
+Project("{9A19103F-16F7-4668-BE54-9A1E7A4F7556}") = "aspnet-mvc", "aspnet-mvc\aspnet-mvc.csproj", "{D169B00F-66E9-4500-879D-487FE0E09F23}"
```
I'm not really a fan of the MVC example, could I convince you to instead use an ASP.NET Core Minimal API instead? It adds a lot less file noise here, and it conveys the same teachings.

noahfalk replied:
Yeah, I felt bad about the large number of files. What if I converted it to inline source in the .md rather than a separate project?

 The challenge I hit with minimal API is that it doesn't naturally lead to using separate classes that will have constructor parameters. I can do method parameter injection instead, but then I don't get unique class names to show as the ILogger\<T> type parameter. Alternately I can create custom classes and register them in a DI container for minimal API, but then it felt like I need to show the extra DI registration and routing hookup or its unclear how the class I am introducing relates whereas MVC controller has a strong convention around it automatically.

noahfalk replied:
> offers a more streamlined solution with fewer components in motion.

Its more streamlined in the size of the full project, but needs more lines of text shown for folks reading the main doc. That isn't how I was prioritizing it before but I'll switch around based on your preference.

noahfalk replied:
switched, mvc is gone and minimal API is added instead.

IEvangelist reviewed (changes requested):
Hi @noahfalk - this looks great, I have some nits and suggestions, as well as several questions.

reyang commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,106 @@ title: Logging in C#
 
-Logging configuration is commonly provided by the `Logging` section of *appsettings*.`{Environment}`*.json* files. The following *appsettings.Development.json* file is generated by the .NET Worker service templates:
+Logging configuration can be set in code or via external sources such as config files and environment variables. Using external configuration is beneficial when possible because
+it can be changed without rebuilding the application. However some tasks, such as setting logging providers, can only be configured from code.
```
Just "without rebuilding the application" or "without rebuilding the application or restarting the process"?

noahfalk replied:
It depends. For environment variables a restart is needed, for config files it isn't. Since I had left the configuration source open-ended I only mentioned no-rebuild which is true for all of them.

reyang commented on docs/core/extensions/logging.md:
```diff
@@ -169,6 +241,34 @@ On [Azure App Service](https://azure.microsoft.com/services/app-service/), selec
+});
+
+ILogger logger = loggerFactory.CreateLogger<Program>();
+logger.LogDebug("Example log message");
```
Minor, can we change this to a message template with one or two key-value pairs? (to avoid readers interpreting it as logging a raw string)

reyang commented on docs/core/extensions/snippets/logging/getting-started-logger-message/Program.cs:
```diff
@@ -0,0 +1,14 @@
+        LogStartupMessage(logger, "fun");
+    }
+
+    [LoggerMessage(Level = LogLevel.Information, Message = "Hello World! Logging is {description}.")]
```
```suggestion
    [LoggerMessage(LogLevel.Information, "Hello World! Logging is {description}.")]
```

noahfalk replied:
I think this one relies on a new constructor that is only present in .NET 8. I left it out since the samples don't (yet) depend on .NET 8

reyang commented on docs/core/extensions/snippets/logging/getting-started-open-telemetry/getting-started-open-telemetry.csproj:
```diff
@@ -0,0 +1,17 @@
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
+    <RootNamespace>getting_started_open_telemetry</RootNamespace>
```
Is this needed? (other projects don't have it)

noahfalk replied:
No it isn't needed, I just forgot to remove it.

IEvangelist commented on docs/core/extensions/snippets/logging/getting-started-logger-message/Program.cs:
```diff
@@ -0,0 +1,14 @@
+    static void Main(string[] args)
+    {
+        using ILoggerFactory factory = LoggerFactory.Create(builder => builder.AddConsole());
+        ILogger logger = factory.CreateLogger("Program");
```
Why not use `factory.CreateLogger<Program>();`?

noahfalk replied:
I was trying to avoid adding another concept I'd need to explain in the introductory sample. As-is we can describe category as a string associated with the ILogger and since the parameter type is a string I expect everyone to understand. If instead we use factory.CreateLogger<Program>() then the explanation of how the type parameter maps to a string has to be front-loaded.

IEvangelist commented on docs/core/extensions/snippets/logging/getting-started-open-telemetry/Program.cs:
```diff
@@ -0,0 +1,12 @@
+    });
+});
+ILogger logger = factory.CreateLogger("Program");
+logger.LogInformation("Hello World! Logging is {description}.", "fun");
```
```suggestion
logger.LogInformation("Hello World! Logging is {Description}.", "fun");
```

noahfalk commented on docs/core/extensions/snippets/logging/di-without-host/di-without-host.csproj:
```diff
@@ -0,0 +1,17 @@
+
+  <PropertyGroup>
+    <OutputType>Exe</OutputType>
+    <TargetFramework>net7.0</TargetFramework>
```
What is our usual stance on using new versions of .NET in samples prior to that version being released? The docs don't currently mention .NET 8 is a pre-requisite or use any features that are specific to .NET 8. 

For primary constructors my personal opinion would be that we should be conservative in the language features we use in samples. Many devs may be unfamiliar with the new features and then it would be harder for them to understand the sample content if those features are used. Of course if you tell me the docs team has considered this in the past and found it isn't an issue I'll trust your expertise.

IEvangelist replied:
That is a fair question. We are slowing adding .NET 8 bits into the fold, since we're so close to .NET 8 going GA - I'd advocate for that. And for introducing new C# language features into content, we strive to do this as much as possible. For the very reasons that you mentioned, many devs are unfamiliar with them and we believe this is the appropriate place to show them how to use these features. The .NET docs team is consistently introducing the new C# syntax throughout our examples.

noahfalk replied:
I added a primary constructor the first place constructor injection comes up. There are other potential places it could be used but I didn't want to get that aggressive with it.

noahfalk committed:
Apply suggestions from code review

Co-authored-by: David Pine <david.pine@microsoft.com>
Co-authored-by: Reiley Yang <reyang@microsoft.com>

noahfalk commented on docs/core/extensions/snippets/logging/aspnet-mvc/aspnet-mvc.csproj:
```diff
@@ -0,0 +1,15 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="OpenTelemetry" Version="1.6.0" />
```
Just the protocol is sufficient, I removed the other reference.

noahfalk commented on docs/core/extensions/snippets/logging/getting-started-open-telemetry/getting-started-open-telemetry.csproj:
```diff
@@ -0,0 +1,17 @@
+  </PropertyGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Logging" Version="7.0.0" />
+    <PackageReference Include="OpenTelemetry" Version="1.6.0" />
+    <PackageReference Include="OpenTelemetry.Exporter.OpenTelemetryProtocol" Version="1.6.0" />
```
Empirically it works with just the OpenTelemetryProtocol package right now. I removed the OpenTelemetry package but left the logging package. I expect it is common that people will have that package reference if they add logging prior to adding OTel as a provider. I don't think we gain any particular value in encouraging people to remove the reference.

noahfalk committed:
Feedback from PR review

noahfalk committed:
Feedback from PR review

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -3,34 +3,91 @@ title: Logging in C#
+The preceding example:
+
+- Created a DI service container containing an `ILoggerFactory` configured to write to the console
+- Added a singleton `Worker` to the container
+- Created an instance of the `Worker` from the DI container which also automatically created an `ILogger<Worker>` to use as the constructor argument.
+- Invoked `Worker.DoSomeWork` which used the `ILogger<Worker>` to log a message to the console.
```
```suggestion
- Added a singleton `ExampleService` to the container
- Created an instance of the `ExampleService` from the DI container which also automatically created an `ILogger<ExampleService>` to use as the constructor argument.
- Invoked `ExampleService.DoSomeWork` which used the `ILogger<ExampleService>` to log a message to the console.
```

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -169,6 +226,34 @@ On [Azure App Service](https://azure.microsoft.com/services/app-service/), selec
+});
+
+ILogger logger = loggerFactory.CreateLogger<Program>();
+logger.LogDebug("Hello {target}", "Everyone");
```
```suggestion
logger.LogDebug("Hello {Target}", "Everyone");
```

noahfalk replied:
sorry I feel silly - after fixing ~5 other instances of that same capitalization issue I added a new one :( Thanks for catching it.

IEvangelist replied:
No worries, all good. 🤓

IEvangelist commented:
I just noticed too, now that we've updated the source a bit - we'll have to fix the `range=""` usage. With code examples that are not fully rendered, i.e.; if we're doing ranges only for a given file, we could instead use `id="some-dentifier"`.

Here's an example:

```cs
// <main>
static void Main(string[] args)
{
    using ILoggerFactory factory = LoggerFactory.Create(builder => builder.AddConsole());
    ILogger logger = factory.CreateLogger("Program");
    LogStartupMessage(logger, "fun");
}
// </main>

// <logger-message>
[LoggerMessage(Level = LogLevel.Information, Message = "Hello World! Logging is {Description}.")]
static partial void LogStartupMessage(ILogger logger, string description);
// </logger-message>
```

We could render each section independently, by using an `id`.

```markdown
:::code source="path/to/file.cs" id="main":::
``` 

This would render as:

```cs
static void Main(string[] args)
{
    using ILoggerFactory factory = LoggerFactory.Create(builder => builder.AddConsole());
    ILogger logger = factory.CreateLogger("Program");
    LogStartupMessage(logger, "fun");
}
```

noahfalk committed:
Updates from PR feedback

noahfalk committed:
Fix lint issues and adjust highlight

noahfalk committed:
switch back to range syntax

noahfalk commented:
> I just noticed too, now that we've updated the source a bit - we'll have to fix the `range=""` usage. With code examples that are not fully rendered, i.e.; if we're doing ranges only for a given file, we could instead use `id="some-dentifier"`.


I attempted the id technique in one spot but got this warning from CI: "Line 70: [Warning] It looks like your code snippet was not rendered. Try range instead." Its possible I did it wrong but I switched it back to range as suggested and I think the snippets are correctly displayed and highlighted now.

noahfalk committed:
grammar fix

noahfalk commented:
Fingers crossed all the feedback has been addressed. Let me know if there is anything else you'd like me to take care of. Thanks!

IEvangelist commented on docs/core/extensions/logging.md:
```diff
@@ -33,51 +33,53 @@ There are several changes you should consider making to the previous example whe
+  - Created a singleton service called `ExampleHandler` and mapped incoming web requests to run the `ExampleHandler.HandleRequest` function.
+  - Line 8 defines a [primary constructor](/dotnet/csharp/whats-new/tutorials/primary-constructors) for the ExampleHandler, a feature added in C# 12. Using the older style C# constructor would work equally well but is a little more verbose.
+  - The constructor defines a parameter of type `ILogger<ExampleHandler>`. <xref:Microsoft.Extensions.Logging.ILogger%601> derives from <xref:Microsoft.Extensions.Logging.ILogger> and indicates which category the `ILogger` object has. The DI container locates an `ILogger` with the correct category and supplies it as the constructor argument. If no `ILogger` with that category exists yet, the DI c ...
+  - The `logger` parameter received in the constructor was used for logging in the `HandleRequest` function.
```
Address MD linter failed checks.

```suggestion
- Created a singleton service called `ExampleHandler` and mapped incoming web requests to run the `ExampleHandler.HandleRequest` function.
- Line 8 defines a [primary constructor](/dotnet/csharp/whats-new/tutorials/primary-constructors) for the ExampleHandler, a feature added in C# 12. Using the older style C# constructor would work equally well but is a little more verbose.
- The constructor defines a parameter of type `ILogger<ExampleHandler>`. <xref:Microsoft.Extensions.Logging.ILogger%601> derives from <xref:Microsoft.Extensions.Logging.ILogger> and indicates which category the `ILogger` object has. The DI container locates an `ILogger` with the correct category and supplies it as the constructor argument. If no `ILogger` with that category exists yet, the DI container automatically creates it from the `ILoggerFactory` in the service provider.
- The `logger` parameter received in the constructor was used for logging in the `HandleRequest` function.
```

IEvangelist commented on docs/core/extensions/snippets/logging/minimal-web-open-telemetry/Program.cs:
Love this! 💜

IEvangelist commented on docs/core/extensions/snippets/logging/minimal-web-open-telemetry/minimal-web-open-telemetry.csproj:
```diff
@@ -0,0 +1,10 @@
+    <TargetFramework>net8.0</TargetFramework>
+    <Nullable>enable</Nullable>
+    <ImplicitUsings>enable</ImplicitUsings>
+    <RootNamespace>minimal_web_open_telemetry</RootNamespace>
```
Not a fan of this namespace.

IEvangelist reviewed (approved):
Looks good to me, we'll :shipit: - thank you 🙏
